lib/internal/per_context/safethenable.js

```diff
@@ -0,0 +1,71 @@
+'use strict';
+
+const {
+  PromisePrototypeCatch,
+  PromisePrototypeThen,
+  ReflectApply,
+  SafeWeakMap,
+} = primordials;
+
+const { isPromise } = require('internal/util/types');
+
+const cache = new SafeWeakMap();
+
+// `SafeThenable` should be used when dealing with user provided `Promise`-like
+// instances. It provides two methods `.then` and `.catch`.
+// Wrapping uses of `SafeThenable` in `try`/`catch` may be useful if the
+// accessing `.then` property of the provided object throws.
+class SafeThenable {
+  #is_promise;
+  #makeUnsafeCalls;
+  #target;
+  #cachedThen;
+  #hasAlreadyAccessedThen;
+
+  constructor(thenable, makeUnsafeCalls = false) {
+    this.#target = thenable;
+    this.#is_promise = isPromise(thenable);
+    this.#makeUnsafeCalls = makeUnsafeCalls;
+  }
+
+  get #then() {
```

Comment:
While I'm fine with using this syntax in general, we should benchmark this extensively before landing. Last I checked, v8 was not yet optimizing private accessors that well and since we're using this in EventEmitter (one of the most performance sensitive bits of code we have in core) it's good to be careful here.